### ISSUE TITLE
podman_container always recreates my container even if mount/volume config does not change

### ISSUE THREAD

**Issue by rubiksdot** [containers/ansible-podman-collections#376]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

Rerunning containers.podman.podman_container with the same settings as used to create the container causes the running container to be stopped, removed and recreated because it thinks that the mount/volume configuration has changed.

**Steps to reproduce the issue:**

1. Run containers.podman.podman_container to create container with mounts configured via "mount" and with state "created" or "present".

2. Re-run containers.podman.podman_container with the exact same settings.

3. Watch your container get nuked and re-created.

**Describe the results you received:**

My created container gets removed and recreated.

**Describe the results you expected:**

Container be left alone as the configuration has not changed.

**Additional information you deem important (e.g. issue happens only occasionally):**

Judging from the code PodmanContainerDiff.diffparam_volume() is involved here and it compares "mounts" in before with "volume" in after. Given that I'm only using mount this results it comparing an existing list of mounts with nothing and so it decides to recreate the container as it thinks the config has changed.

As near as I can tell it should either merge the mount and volume lists in before and after so that the check compares everything or compare mounts with mounts and volumes with volumes if they should not be mixed. The current way either compares incomplete lists or apples with oranges.

**Version of the `containers.podman` collection:**
**Either git commit if installed from git: `git show --summary`**
**Or version from `ansible-galaxy` if installed from galaxy: `ansible-galaxy collection list | grep containers.podman`**

```
containers.podman         1.4.1  
containers.podman             1.5.0  
containers.podman 1.9.1  
```

**Output of `ansible --version`:**

```
ansible 2.10.17
  config file = /data/ansible/ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /root/.local/lib/python3.9/site-packages/ansible
  executable location = /root/.local/bin/ansible
  python version = 3.9.2 (default, Feb 28 2021, 17:03:44) [GCC 10.2.1 20210110]
```

**Output of `podman version`:**

```
Version:      3.0.1
API Version:  3.0.0
Go Version:   go1.15.9
Built:        Thu Jan  1 10:00:00 1970
OS/Arch:      linux/amd64
```

**Output of `podman info --debug`:**

``` yaml
host:
  arch: amd64
  buildahVersion: 1.19.6
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: 'conmon: /usr/bin/conmon'
    path: /usr/bin/conmon
    version: 'conmon version 2.0.25, commit: unknown'
  cpus: 1
  distribution:
    distribution: debian
    version: "11"
  eventLogger: journald
  hostname: xxx
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 5.15.0-0.bpo.2-amd64
  linkmode: dynamic
  memFree: 130334720
  memTotal: 1022939136
  ociRuntime:
    name: crun
    package: 'crun: /usr/bin/crun'
    path: /usr/bin/crun
    version: |-
      crun version 0.17
      commit: 0e9229ae34caaebcb86f1fde18de3acaf18c6d9a
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
  os: linux
  remoteSocket:
    exists: true
    path: /run/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: false
    seccompEnabled: true
    selinuxEnabled: false
  slirp4netns:
    executable: ""
    package: ""
    version: ""
  swapFree: 243535872
  swapTotal: 268431360
  uptime: 73h 13m 14.22s (Approximately 3.04 days)
registries: {}
store:
  configFile: /etc/containers/storage.conf     
  containerStore:
    number: 1
    paused: 0
    running: 1
    stopped: 0
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /var/lib/containers/storage
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "true"
  imageStore:
    number: 28
  runRoot: /run/containers/storage
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 3.0.0
  Built: 0
  BuiltTime: Thu Jan  1 10:00:00 1970
  GitCommit: ""
  GoVersion: go1.15.9
  OsArch: linux/amd64
  Version: 3.0.1
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman/stable,stable,now 3.0.1+dfsg1-3+b2 amd64 [installed]
```

**Playbok you run with ansible (e.g. content of `playbook.yaml`):**

``` yaml
# vim:ts=2:et:ft=yaml

---
# vim:ts=2:et:ft=yaml

---
- hosts: xxx

  vars:
    pod_name: lighty-debrepo

  tasks:
  - name: "{{pod_name}} : create container"
    containers.podman.podman_container:
      state: present
      cap_drop: "cap_mknod"
      detach: true
      expose: 80
      generate_systemd: {
        path: /etc/systemd/system
      }
      #hostname:
      #image:
      ip: "10.200.1.2"
      mac_address: "ee:ac:59:57:4c:af"
      network: moocow
      log_driver: journald
      log_level: info
      mount: [
        "type=bind,src=/home/podman/{{create_package_item.name}},dst=/mnt",
      ]
      tmpfs: {
        "/tmp":     "exec",
        "/var/tmp": "exec",
        "/run":     "rw,noexec,nosuid,nodev",
      }
      image: "{{create_package_item.name}}"
      name: "{{create_package_item.name}}"
      read_only: true
      #read_only_tmpfs: true
    loop:
      - { name: "debian-amd64-{{pod_name}}" }
    loop_control:
      loop_var: create_package_item
```

**Command line and output of ansible run with high verbosity**

**Please NOTE: if you submit a bug about idempotency, run the playbook with `--diff` option, like:**

`ansible-playbook -i inventory --diff -vv playbook.yml`

```
ansible-playbook 2.10.17
  config file = /data/ansible/ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /root/.local/lib/python3.9/site-packages/ansible
  executable location = /root/.local/bin/ansible-playbook
  python version = 3.9.2 (default, Feb 28 2021, 17:03:44) [GCC 10.2.1 20210110]
Using /data/ansible/ansible.cfg as config file
Skipping callback 'default', as we already have a stdout callback.
Skipping callback 'minimal', as we already have a stdout callback.
Skipping callback 'oneline', as we already have a stdout callback.

PLAYBOOK: meh.yml *****************************************************************************************************************************************************************************
1 plays in meh.yml

PLAY [xxx] **********************************************************************************************************************************************************************************

TASK [Gathering Facts] ************************************************************************************************************************************************************************
task path: /data/ansible/meh.yml:4
ok: [xxx]
META: ran handlers

TASK [lighty-debrepo : create container] ******************************************************************************************************************************************************
task path: /data/ansible/meh.yml:10
changed: [xxx] => (item={'name': 'debian-amd64-lighty-debrepo'}) => {"actions": ["recreated debian-amd64-lighty-debrepo"], "ansible_loop_var": "create_package_item", "changed": true, "container": {"AppArmorProfile": "", "Args": ["3600"], "BoundingCaps": ["CAP_CHOWN", "CAP_DAC_OVERRIDE", "CAP_FOWNER", "CAP_FSETID", "CAP_KILL", "CAP_NET_BIND_SERVICE", "CAP_SETFCAP", "CAP_SETGID", "CAP_SETPCAP", "CAP_SETUID", "CAP_SYS_CHROOT"], "Config": {"Annotations": {"io.kubernetes.cri-o.TTY": "false", "io.podman.annotations.autoremove": "FALSE", "io.podman.annotations.init": "FALSE", "io.podman.annotations.privileged": "FALSE", "io.podman.annotations.publish-all": "FALSE"}, "AttachStderr": false, "AttachStdin": false, "AttachStdout": false, "Cmd": ["sleep", "3600"], "CreateCommand": ["podman", "container", "create", "--name", "debian-amd64-lighty-debrepo", "--cap-drop", "cap_mknod", "--expose", "80", "--ip", "10.200.1.2", "--mac-address", "ee:ac:59:57:4c:af", "--network", "moocow", "--log-driver", "journald", "--log-level", "info", "--mount", "type=bind,src=/home/podland/debian-amd64-lighty-debrepo,dst=/mnt", "--tmpfs", "/tmp:exec", "--tmpfs", "/var/tmp:exec", "--tmpfs", "/run:rw,noexec,nosuid,nodev", "--read-only=True", "debian-amd64-lighty-debrepo"], "Domainname": "", "Entrypoint": "", "Env": ["PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin", "TERM=xterm", "container=podman"], "Hostname": "0af819119125", "Image": "localhost/debian-amd64-lighty-debrepo:latest", "Labels": {"io.buildah.version": "1.19.6"}, "OnBuild": null, "OpenStdin": false, "StdinOnce": false, "StopSignal": 15, "Tty": false, "Umask": "0022", "User": "", "Volumes": null, "WorkingDir": "/root/"}, "ConmonPidFile": "/run/containers/storage/overlay-containers/0af819119125815a12db86a1a0104df414c26ee349504ab187dcaebc513c3121/userdata/conmon.pid", "Created": "2022-02-17T21:35:34.260495687+11:00", "Dependencies": [], "Driver": "overlay", "EffectiveCaps": ["CAP_CHOWN", "CAP_DAC_OVERRIDE", "CAP_FOWNER", "CAP_FSETID", "CAP_KILL", "CAP_NET_BIND_SERVICE", "CAP_SETFCAP", "CAP_SETGID", "CAP_SETPCAP", "CAP_SETUID", "CAP_SYS_CHROOT"], "ExecIDs": [], "ExitCommand": ["/usr/bin/podman", "--root", "/var/lib/containers/storage", "--runroot", "/run/containers/storage", "--log-level", "info", "--cgroup-manager", "systemd", "--tmpdir", "/run/libpod", "--runtime", "crun", "--events-backend", "journald", "container", "cleanup", "0af819119125815a12db86a1a0104df414c26ee349504ab187dcaebc513c3121"], "GraphDriver": {"Data": {"LowerDir": "/var/lib/containers/storage/overlay/d8defb968674f06f308648d39d75276cc6a089cd9ffe4ed749f73326108cc84b/diff:/var/lib/containers/storage/overlay/2cb79a5c381a5efcfe437c592a14094de1bd54873feb410e7557a141c90c9164/diff:/var/lib/containers/storage/overlay/089be3984c07c83e4eae3f2bd1c53128edbf630b3d3914e2786d98aaa7e8e96d/diff:/var/lib/containers/storage/overlay/e4046281b54d0667fc5791cf39372e1543487c6b8adb9ba7615da7a036950986/diff:/var/lib/containers/storage/overlay/3a8c0d637de0be5a5e5658700630ac2ea4cf6aedf765f6360cab65af708615a1/diff:/var/lib/containers/storage/overlay/d2d7ce71f7623dc35ff01f2f5ffd37f45eb64b59cad2a2089e28f1d546a08038/diff:/var/lib/containers/storage/overlay/2c64aed5f3adbe12633b6efb91d36f70a6ff6ba35f96dc09803519a6f2d713b1/diff:/var/lib/containers/storage/overlay/a16f0ed7f6ccac4a200835c3bc845252d223d0b1471096fc251a3c35b4b97548/diff:/var/lib/containers/storage/overlay/20dd14f721cecc69507cde03186a55cac40e8cd0bc172bd0445674f1226db417/diff:/var/lib/containers/storage/overlay/98289348f737a05a6351e576c9fef16790ab1e24a455defc8ec0de650761c863/diff:/var/lib/containers/storage/overlay/035b5aadc71336f859fac40d3af0bfbc79d9244a60536f205c346e2d9cdf9b61/diff:/var/lib/containers/storage/overlay/46d40a79d4fbb09bae5376a24f423f99ff5eb2aa6c4ff667235fa30b068a0c47/diff:/var/lib/containers/storage/overlay/99480cae04773c359246eba68724b67f534a15ffb608d3e020fdc04f586d8558/diff:/var/lib/containers/storage/overlay/fdf475cd197fa58549d21dc23fa53999018ae7a9575cd0177033378dcd794648/diff:/var/lib/containers/storage/overlay/803ffdff81e559a8910e455257349bd81e707d3f9aee24f1c5f24d1356730b31/diff:/var/lib/containers/storage/overlay/7cc10dec0e7a849cbd946ffbb6957faa50b18957305a917d52f47e3578857514/diff", "UpperDir": "/var/lib/containers/storage/overlay/24b4af43c548f5a8b5ae6672d0826910c5addc5296d825081ef17409a61cdb96/diff", "WorkDir": "/var/lib/containers/storage/overlay/24b4af43c548f5a8b5ae6672d0826910c5addc5296d825081ef17409a61cdb96/work"}, "Name": "overlay"}, "HostConfig": {"AutoRemove": false, "Binds": ["/home/podland/debian-amd64-lighty-debrepo:/mnt:rw,rprivate,rbind"], "BlkioDeviceReadBps": null, "BlkioDeviceReadIOps": null, "BlkioDeviceWriteBps": null, "BlkioDeviceWriteIOps": null, "BlkioWeight": 0, "BlkioWeightDevice": null, "CapAdd": [], "CapDrop": ["CAP_MKNOD", "CAP_NET_RAW", "CAP_AUDIT_WRITE"], "Cgroup": "", "CgroupConf": null, "CgroupManager": "systemd", "CgroupMode": "private", "CgroupParent": "", "Cgroups": "default", "ConsoleSize": [0, 0], "ContainerIDFile": "", "CpuCount": 0, "CpuPercent": 0, "CpuPeriod": 0, "CpuQuota": 0, "CpuRealtimePeriod": 0, "CpuRealtimeRuntime": 0, "CpuShares": 0, "CpusetCpus": "", "CpusetMems": "", "Devices": [], "DiskQuota": 0, "Dns": [], "DnsOptions": [], "DnsSearch": [], "ExtraHosts": [], "GroupAdd": [], "IOMaximumBandwidth": 0, "IOMaximumIOps": 0, "IpcMode": "private", "Isolation": "", "KernelMemory": 0, "Links": null, "LogConfig": {"Config": null, "Path": "", "Size": "0B", "Tag": "", "Type": "journald"}, "Memory": 0, "MemoryReservation": 0, "MemorySwap": 0, "MemorySwappiness": 0, "NanoCpus": 0, "NetworkMode": "bridge", "OomKillDisable": false, "OomScoreAdj": 0, "PidMode": "private", "PidsLimit": 2048, "PortBindings": {}, "Privileged": false, "PublishAllPorts": false, "ReadonlyRootfs": true, "RestartPolicy": {"MaximumRetryCount": 0, "Name": ""}, "Runtime": "oci", "SecurityOpt": [], "ShmSize": 65536000, "Tmpfs": {"/run": "rw,noexec,nosuid,nodev,rprivate,tmpcopyup", "/tmp": "exec,rw,rprivate,nosuid,nodev,tmpcopyup", "/var/tmp": "exec,rw,rprivate,nosuid,nodev,tmpcopyup"}, "UTSMode": "private", "Ulimits": [{"Hard": 1048576, "Name": "RLIMIT_NOFILE", "Soft": 1048576}, {"Hard": 4194304, "Name": "RLIMIT_NPROC", "Soft": 4194304}], "UsernsMode": "", "VolumeDriver": "", "VolumesFrom": null}, "HostnamePath": "", "HostsPath": "", "Id": "0af819119125815a12db86a1a0104df414c26ee349504ab187dcaebc513c3121", "Image": "11741c5e21cf69d742417750ffde0665369888cf58f395d21f07f1e7dd950f7b", "ImageName": "localhost/debian-amd64-lighty-debrepo:latest", "IsInfra": false, "MountLabel": "", "Mounts": [{"Destination": "/mnt", "Driver": "", "Mode": "", "Options": ["rbind"], "Propagation": "rprivate", "RW": true, "Source": "/home/podland/debian-amd64-lighty-debrepo", "Type": "bind"}], "Name": "debian-amd64-lighty-debrepo", "Namespace": "", "NetworkSettings": {"Bridge": "", "EndpointID": "", "Gateway": "", "GlobalIPv6Address": "", "GlobalIPv6PrefixLen": 0, "HairpinMode": false, "IPAddress": "", "IPPrefixLen": 0, "IPv6Gateway": "", "LinkLocalIPv6Address": "", "LinkLocalIPv6PrefixLen": 0, "MacAddress": "", "Networks": {"moocow": {"DriverOpts": null, "EndpointID": "", "Gateway": "", "GlobalIPv6Address": "", "GlobalIPv6PrefixLen": 0, "IPAMConfig": null, "IPAddress": "", "IPPrefixLen": 0, "IPv6Gateway": "", "Links": null, "MacAddress": "", "NetworkID": "moocow"}}, "Ports": {}, "SandboxID": "", "SandboxKey": ""}, "OCIRuntime": "crun", "Path": "sleep", "Pod": "", "ProcessLabel": "", "ResolvConfPath": "", "RestartCount": 0, "Rootfs": "", "State": {"Dead": false, "Error": "", "ExitCode": 0, "FinishedAt": "0001-01-01T00:00:00Z", "Healthcheck": {"FailingStreak": 0, "Log": null, "Status": ""}, "OOMKilled": false, "OciVersion": "1.0.2-dev", "Paused": false, "Pid": 0, "Restarting": false, "Running": false, "StartedAt": "0001-01-01T00:00:00Z", "Status": "configured"}, "StaticDir": "/var/lib/containers/storage/overlay-containers/0af819119125815a12db86a1a0104df414c26ee349504ab187dcaebc513c3121/userdata"}, "create_package_item": {"name": "debian-amd64-lighty-debrepo"}, "podman_actions": ["podman rm -f debian-amd64-lighty-debrepo", "podman create --name debian-amd64-lighty-debrepo --cap-drop cap_mknod --expose 80 --ip 10.200.1.2 --mac-address ee:ac:59:57:4c:af --network moocow --log-driver journald --log-level info --mount type=bind,src=/home/podland/debian-amd64-lighty-debrepo,dst=/mnt --tmpfs /tmp:exec --tmpfs /var/tmp:exec --tmpfs /run:rw,noexec,nosuid,nodev --read-only=True debian-amd64-lighty-debrepo"], "podman_systemd": {"container-debian-amd64-lighty-debrepo": "# container-debian-amd64-lighty-debrepo.service\n# autogenerated by Podman 3.0.1\n# Thu Feb 17 21:35:34 AEDT 2022\n\n[Unit]\nDescription=Podman container-debian-amd64-lighty-debrepo.service\nDocumentation=man:podman-generate-systemd(1)\nWants=network.target\nAfter=network-online.target\n\n[Service]\nEnvironment=PODMAN_SYSTEMD_UNIT=%n\nRestart=on-failure\nTimeoutStopSec=70\nExecStart=/usr/bin/podman start debian-amd64-lighty-debrepo\nExecStop=/usr/bin/podman stop -t 10 debian-amd64-lighty-debrepo\nExecStopPost=/usr/bin/podman stop -t 10 debian-amd64-lighty-debrepo\nPIDFile=/run/containers/storage/overlay-containers/0af819119125815a12db86a1a0104df414c26ee349504ab187dcaebc513c3121/userdata/conmon.pid\nType=forking\n\n[Install]\nWantedBy=multi-user.target default.target\n"}, "stderr": "time=\"2022-02-17T21:35:34+11:00\" level=info msg=\"podman filtering at log level info\"\ntime=\"2022-02-17T21:35:34+11:00\" level=info msg=\"Not using native diff for overlay, this may cause degraded performance for building images: kernel has CONFIG_OVERLAY_FS_REDIRECT_DIR enabled\"\ntime=\"2022-02-17T21:35:34+11:00\" level=info msg=\"[graphdriver] using prior storage driver: overlay\"\ntime=\"2022-02-17T21:35:34+11:00\" level=info msg=\"Error initializing configured OCI runtime kata: no valid executable found for OCI runtime kata: invalid argument\"\ntime=\"2022-02-17T21:35:34+11:00\" level=info msg=\"Error initializing configured OCI runtime runc: no valid executable found for OCI runtime runc: invalid argument\"\ntime=\"2022-02-17T21:35:34+11:00\" level=info msg=\"Found CNI network podman (type=ptp) at /etc/cni/net.d/87-podman-ptp.conflist\"\ntime=\"2022-02-17T21:35:34+11:00\" level=info msg=\"Found CNI network moocow (type=bridge) at /etc/cni/net.d/moocow.conflist\"\ntime=\"2022-02-17T21:35:34+11:00\" level=info msg=\"Setting parallel job count to 4\"\n", "stderr_lines": ["time=\"2022-02-17T21:35:34+11:00\" level=info msg=\"podman filtering at log level info\"", "time=\"2022-02-17T21:35:34+11:00\" level=info msg=\"Not using native diff for overlay, this may cause degraded performance for building images: kernel has CONFIG_OVERLAY_FS_REDIRECT_DIR enabled\"", "time=\"2022-02-17T21:35:34+11:00\" level=info msg=\"[graphdriver] using prior storage driver: overlay\"", "time=\"2022-02-17T21:35:34+11:00\" level=info msg=\"Error initializing configured OCI runtime kata: no valid executable found for OCI runtime kata: invalid argument\"", "time=\"2022-02-17T21:35:34+11:00\" level=info msg=\"Error initializing configured OCI runtime runc: no valid executable found for OCI runtime runc: invalid argument\"", "time=\"2022-02-17T21:35:34+11:00\" level=info msg=\"Found CNI network podman (type=ptp) at /etc/cni/net.d/87-podman-ptp.conflist\"", "time=\"2022-02-17T21:35:34+11:00\" level=info msg=\"Found CNI network moocow (type=bridge) at /etc/cni/net.d/moocow.conflist\"", "time=\"2022-02-17T21:35:34+11:00\" level=info msg=\"Setting parallel job count to 4\""], "stdout": "0af819119125815a12db86a1a0104df414c26ee349504ab187dcaebc513c3121\n", "stdout_lines": ["0af819119125815a12db86a1a0104df414c26ee349504ab187dcaebc513c3121"]}
META: ran handlers
META: ran handlers

PLAY RECAP ************************************************************************************************************************************************************************************
xxx                      : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

KVM host

**Comment by sshnaidm** [comment]
Seems same as #355

**Comment by sshnaidm** [comment]
Fixed in #750
